src/to-be-visible.js

```diff
@@ -1,33 +1,63 @@
-  return (
-    display !== 'none' &&
-    visibility !== 'hidden' &&
-    visibility !== 'collapse' &&
-    opacity !== '0' &&
-    opacity !== 0
-  )
+function isVisibleSummaryDetails(element, previousElement) {
+  return element.nodeName === 'DETAILS' &&
+    previousElement.nodeName !== 'SUMMARY'
+    ? element.hasAttribute('open')
+    : true
 }
 
-function isAttributeVisible(element, previousElement) {
+function isElementTreeVisible(element, previousElement = undefined) {
+  const {getComputedStyle} = element.ownerDocument.defaultView
+  const {display, opacity} = getComputedStyle(element)
   return (
+    display !== 'none' &&
+    opacity !== '0' &&
     !element.hasAttribute('hidden') &&
-    (element.nodeName === 'DETAILS' && previousElement.nodeName !== 'SUMMARY'
-      ? element.hasAttribute('open')
+    isVisibleSummaryDetails(element, previousElement) &&
+    (element.parentElement
+      ? isElementTreeVisible(element.parentElement, element)
       : true)
   )
 }
 
-function isElementVisible(element, previousElement) {
-  return (
-    isStyleVisible(element) &&
-    isAttributeVisible(element, previousElement) &&
-    (!element.parentElement || isElementVisible(element.parentElement, element))
-  )
+function isElementVisibilityVisible(element) {
+  const visibility = getElementVisibilityStyle(element)
+  return visibility !== 'hidden' && visibility !== 'collapse'
+}
+
+/**
+ * Computes the boolean value that determines if an element is considered visible from the
+ * `toBeVisible` custom matcher point of view.
+ *
+ * Visibility is controlled via two different sets of properties and styles.
+ *
+ * 1. One set of properties allow parent elements to fully controls its sub-tree visibility. This
+ *    means that if higher up in the tree some element is not visible by this criteria, it makes the
+ *    entire sub-tree not visible too, and there's nothing that child elements can do to revert it.
+ *    This includes `display: none`, `opacity: 0`, the presence of the `hidden` attribute`, and the
+ *    open state of a details/summary elements pair.
+ *
+ * 2. The other aspect influencing if an element is visible is the CSS `visibility` style. This one
+ *    is also inherited. But unlike the previous case, this one can be reverted by child elements.
+ *    A parent element can set its visibiilty to `hidden` or `collapse`, but a child element setting
+ *    its own styles to `visibility: visible` can rever that, and it makes itself visible. Hence,
+ *    this criteria needs to be checked independently of the other one.
+ *
+ * Hence, the apprach taken by this function is two-fold: it first gets the first set of criteria
+ * out of the way, analyzing the target element and up its tree. If this branch yields that the
+ * element is not visible, there's nothing the element could be doing to revert that, so it returns
+ * false. Only if the first check is true, if proceeds to analyze the `visibility` CSS.
```

Comment:
That's an interesting trade-off. You're basically assuming that checking `visibility` is more expensive than walking up the tree and checking `display` etc on every parent.

How was this trade-off motivated? 

In the other Testing Library implementations we're checking `visibility` on the element first since that allows us to potentially bail out of walking up the tree i.e.
```diff
 function isElementVisible(element) {
-  return isElementTreeVisible(element) && isElementVisibilityVisible(element)
+  return isElementVisibilityVisible(element) && isElementTreeVisible(element)
 }
```

Reply:
I was honestly not thinking of performance, and the comment talks more about the logic we follow. Both branches walk up the tree, and either of these branches could bail out more quickly than the other. There's no way to tell.

Though the check for visibility is simpler, so maybe you're right that on average it could be faster. I'm ok with making this change.